### PR TITLE
improve typing annotations in CachedCall

### PR DESCRIPTION
**Pull request by richvdh**
tighten up some of the typing in CachedCall, which is going to be needed when
Twisted 21.7 brings better typing on Deferred.

### CODE REVIEW

File: synapse/util/caches/cached_call.py [hunk 1]
```diff
@@ -92,13 +98,19 @@ def got_result(r):
...
-        d.callback(self._result)
-        return await d
+        if isinstance(self._result, _Sentinel):
+            await make_deferred_yieldable(self._deferred)
+            assert not isinstance(self._result, _Sentinel)
+
+        if isinstance(self._result, Failure):
+            # I *think* awaiting a failed Deferred is the easiest way to correctly raise
+            # the right exception.
+            d = defer.fail(self._result)
```
Comment (erikjohnston) [inline synapse/util/caches/cached_call.py:108]: Can we not use `Failure.raiseException`?

Comment (richvdh) [reply synapse/util/caches/cached_call.py:108]: yes :facepalm:

File: synapse/util/caches/cached_call.py [hunk 2]
```diff
@@ -65,7 +69,7 @@ def __init__(self, f: Callable[[], Awaitable[TV]]):
         """
         self._callable: Optional[Callable[[], Awaitable[TV]]] = f
         self._deferred: Optional[Deferred] = None
-        self._result: Union[None, Failure, TV] = None
+        self._result: Union[_Sentinel, TV, Failure] = _Sentinel.sentinel
```
Comment (babolivier) [inline synapse/util/caches/cached_call.py:72]: Looks like `_result` is an `Union[_Sentinel, TV, Failure]` but its value is an `object`? Or am I misunderstanding something here?

Comment (richvdh) [reply synapse/util/caches/cached_call.py:72]: I think you're missing something. Or I don't understand the question.

`_result` starts out as `_Sentinel.sentinel`, which (via the magic of Enums) is an instance of `_Sentinel`. It is later set to either a `TV` or a `Failure`.

**Review by clokep — APPROVED** [review]
Looks reasonable to me.

I think some of this could be replaced with non-callback logic, but maybe that's out of scope for improving type annotations. 😄

**Pull request merged**